Sources/BrowserServicesKit/SecureVault/SecureVaultManager.swift

```diff
@@ -142,17 +152,94 @@ extension SecureVaultManager: AutofillSecureVaultDelegate {
                                    didRequestIdentityWithId identityId: Int64,
                                    completionHandler: @escaping (SecureVaultModels.Identity?) -> Void) {
         do {
-            completionHandler(try SecureVaultFactory.default.makeVault(errorReporter: self.delegate)
-                                .identityFor(id: identityId))
+            let vault = try self.vault ?? SecureVaultFactory.default.makeVault(errorReporter: self.delegate)
+            completionHandler(try vault.identityFor(id: identityId))
+
             delegate?.secureVaultManager(self, didAutofill: .identity, withObjectId: identityId)
         } catch {
             os_log(.error, "Error requesting identity: %{public}@", error.localizedDescription)
             completionHandler(nil)
         }
     }
     
+    /// Stores autogenerated credentials sent by the AutofillUserScript, or updates an existing row in the database if credentials already exist.
+    /// The Secure Vault only stores one generated password for a domain, which is updated any time the user selects a new generated password.
+    func storeOrUpdateAutogeneratedCredentials(domain: String, autofillData: AutofillUserScript.DetectedAutofillData) throws {
+        guard autofillData.hasAutogeneratedPassword,
+              let autogeneratedCredentials = autofillData.credentials,
+              !(autogeneratedCredentials.username?.isEmpty ?? true),
+              let passwordData = autogeneratedCredentials.password.data(using: .utf8) else {
+            os_log("Did not meet conditions for silently saving autogenerated credentials, returning early", log: .passwordManager)
+            return
+        }
+
+        let vault = try self.vault ?? SecureVaultFactory.default.makeVault(errorReporter: self.delegate)
+        let accounts = try vault.accountsFor(domain: domain)
+        
+        // As a precaution, check whether an account exists with the matching generated password _and_ a non-nil username.
+        // If so, then the user must have already saved the generated credentials and set a username.
+        
+        for account in accounts where !account.username.isEmpty {
+            if let accountID = account.id,
+               let credentialsForAccount = try vault.websiteCredentialsFor(accountId: accountID),
+               credentialsForAccount.password == passwordData,
+               account.username == autogeneratedCredentials.username ?? "" {
+                os_log("Tried to save autogenerated password but it already exists; not saving credentials", log: .passwordManager)
+                return
```

Comment:
Nit: potentially controversial indentation

Reply:
Oh yeah this is no good, let me fix